collector/zfs.go

```diff
@@ -53,22 +43,44 @@ func init() {
 }
 
 type zfsCollector struct {
-	zfsMetrics []zfsMetric
+	zfsMetrics        []zfsMetric
+	linuxProcpathBase string
+	linuxPathMap      map[string]string
 }
 
 func NewZFSCollector() (Collector, error) {
-	return &zfsCollector{}, nil
+	var z zfsCollector
+	z.linuxProcpathBase = "spl/kstat/zfs"
+	z.linuxPathMap = map[string]string{
+		"zfsArc":       "arcstats",
+		"zfsDmuTx":     "dmu_tx",
+		"zfsFm":        "fm",
+		"zfsFetch":     "zfetchstats",
+		"zfsVdevCache": "vdev_cache_stats",
+		"zfsXuio":      "xuio_stats",
+		"zfsZil":       "zil",
+	}
+	return &z, nil
 }
 
 func (c *zfsCollector) Update(ch chan<- prometheus.Metric) (err error) {
-	// Arcstats
-	err = c.updateArcstats(ch)
-	switch {
-	case err == zfsNotAvailableError:
-		log.Debug(err)
-		return nil
-	case err != nil:
-		return err
+	for _, subsystem := range []string{
```

Comment:
Since you're storing the subsystem names as the keys in `c.linuxPathMap` already, how about:

```go
for subsystem := range c.linuxPathMap {
    err = c.updateZfsStats(subsystem, ch)
}
```